improves performance of VirtualFolder.getAllFiles()

pvojtechovsky opened the pull request:
Contains
1. test which measures performance of VirtualFolder.getAllFiles() 
2. test which measures performance of VirtualFolder.getSubFolders() 
3. code which improves that performance.

The performance improvement consists of these changes:
1. FileSystemFile calls slow File.getCanonicalFile() only once in constructor. It is not called later. I hope it breaks nothing. Please check it.
2. FileSystemFolder calls slow File.getCanonicalFile() only once in constructor. It is not called later. I hope it breaks nothing. Please check it.
3.  VirtualFolder getAllFiles() and getSubFolders() uses helper HashSet to collect distinct set of files

Note: Before the test needed 80s on 900 files. After performance fix it needs 0.2s
Note: The usage of helper HashSet would be enough to fix performance of VirtualFolder getAllFiles() and getSubFolders(), but FileSystemFile.toFile() and FileSystemFolder.toFile() and getPath() is probably called on other places in Spoon too, therefore I suggest to take both changes.

monperrus commented on src/main/java/spoon/support/compiler/VirtualFolder.java:
```diff
@@ -46,10 +46,10 @@ public void addFolder(SpoonFolder o) {
+		Set<SpoonFile> distinctSet = new HashSet<>();
 		for (SpoonFile f : getFiles()) {
 			// we take care not to add a file that was already found in a folder
-			if (!result.contains(f)) {
```
are your sure hashCode/equals of SpoonFile are correct? I'm not sure they are well tested.

pvojtechovsky replied:
The VirtualFile and ZipFile really miss hashCode, so they would not work correctly with the HashSet. Do you agree to add hashCode method to them?

monperrus replied:
I would prefer not to have equals/hashcode in those classes, in order to not to maintain them in the future. And consequently, it seems betetr to avoid a hashset here.

pvojtechovsky replied:
The not using HashSet means a performance problems with bigger number of files. The MS Windows file.equals() method is much faster then File.getCannonicalFile, but it is still quite slow. So recursive calling of it in List.contains needs time too. I will measure it tomorrow and give you exact number.

monperrus commented on src/main/java/spoon/support/compiler/VirtualFolder.java:
```diff
@@ -87,9 +87,10 @@ public SpoonFolder getParent() {
+		Set<SpoonFolder> distinctSet = new HashSet<>();
 		for (SpoonFile f : getAllFiles()) {
 			SpoonFolder folder = f.getParent();
-			if (!result.contains(folder)) {
```
are your sure hashCode/equals of SpoonFolder are correct? I'm not sure they are well tested.

pvojtechovsky replied:
The VirtualFolder really miss hashCode, so it would not work correctly with the HashSet. Do you agree to add hashCode method to it?

monperrus commented on src/test/java/spoon/test/support/ResourceTest.java:
```diff
@@ -90,4 +90,31 @@ public void testFilteringFolder() throws Exception {
+		//Measure how long it needs to return all files
+		long startTime = System.currentTimeMillis();
+		folder.getAllFiles();
+		assertTrue("The VirtualFolder.getAllFiles() is too slow", (System.currentTimeMillis() - startTime)<1000);
```
In Spoon, we try not to make assertions based on currentTimeMillis, they are too dependent on the platform. To me, you can remove the tests. Everybody agrees?

pvojtechovsky replied:
Do you check the times needed by execution of each test? I do not know your testing environment, but some tools reports a warning/error even when test passes, but needs much longer time then previous run of the same test. If you have such execution time comparator, then it would make sense to keep the test there - I would just remove assert which checks absolute time.

pvojtechovsky commented:
I have measured the performance of solution with and without hashSet and the results are similar. So I have rolled back changes in VirtualFolder so the hashSet is not used, so the hashCode methods does not have to be added.
The remaining changes makes sense. They really improves performance 60s->0.2s on 900 files on MS Windows 10. So it makes sense to accept it.
I have removed platform specific test, which failed on bad performance.
I have squashed all commits into one. 
Is there anything else needed to accept this PR?

monperrus commented:
What if we simply remove all usages of getCanonicalFile? Do we get a performance improvement as well?

pvojtechovsky commented:
Yes, it improves performance too, but then two files, which are same will may be return false on File.equals() when they are using different path ... 
/root/aDir/../backInRoot/x.txt
/root/backInRoot/x.txt

monperrus commented:
we can keep the canonical version. However, I'd remove the isCanonical boolean, if not canonicalizable, we throw a SpoonException, for sake of least astonishment.
